systemtest/benchtest/main.go

```diff
@@ -239,13 +239,15 @@ func Run(allBenchmarks ...BenchmarkFunc) error {
 }
 
 func warmup(agents int, events uint, url, token string) error {
-	// Assume a base ingest rate of at least 2000 per second, and dynamically
+	// Assume a base ingest rate of at least 1000 per second, and dynamically
 	// set the context timeout based on this ingest rate, or if lower, default
-	// to 15 seconds. The default 5000 / 2000 ~= 2, so the default 15 seconds
+	// to 15 seconds. The default 5000 / 1000 ~= 5, so the default 15 seconds
 	// will be used instead. Additionally, the number of agents is also taken
 	// into account, since each of the agents will send the number of specified
 	// events to the APM Server, increasing its load.
-	timeout := warmupTimeout(2000, events, maxEPM, agents)
+	// Also account the GOMAXPROCS setting, since it will dictate the goroutines
+	// that are scheduled at any time.
+	timeout := warmupTimeout(1000, events, maxEPM, agents, runtime.GOMAXPROCS(0))
```

Comment:
I'm starting to wonder if this logic is just too complicated and we may be better off limiting the maximum number of agents we warm up with?

Reply:
My hypothesis here is that the warmup timeout calculation is overlooking the overhead of having a high number of agents and If the hypothesis is correct I don't think we need to take the GOMAXPROCS into account. Explanation:

`timeout = math.Max(events/rate*agents, 15) seconds`

Based on the above formula, timeout is proportional to the number of agents as well as number of events. But the math.Max introduces a lower bound of 15 seconds which would be applicable for smaller events and larger agent count. For example, at one extreme, having 1 event and 20000 agents will also give us 15 seconds timeout which will imply that apm-server has to process 20000 requests of 1 event each in 15 seconds.

We can test the hypothesis by fixing the agents to a high number and using lower and higher value events. I tested it with 80 agents and 1 event vs 80 agents and 1000 events. The first one failed with deadline exceeded (calculated timeout = 15 seconds) while the latter passed (calculated timeout = 40 seconds).

One way to address this (assuming the hypothesis is correct) would be to assign a higher weight to the agents.